text/0037-node-upgrades.md

```diff
@@ -0,0 +1,98 @@
+# Node Upgrades
+
+## Summary
+
+[summary]: #summary
+
+CEP PR: [casperlabs/ceps#0037](https://github.com/casperlabs/ceps/pull/0037)
+
+This CEP describes the how the `casper-node-launcher` and `casper-node` will interact to achieve network upgrades.
+
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+The `casper-node-launcher` is a relatively simple Rust binary which runs the `casper-node` binary as a child process.
+Normal operation will see the launcher running indefinitely, cycling upwards through versions of the node as and when
+they become active.
+
+The node has two relevant modes of operation: "validator mode", and "migrate-data mode".  The normal mode is validator.
+The migrate-data mode is used immediately after an upgrade to migrate any data stored on disk as required.
+
+The node also has three possible exit codes: 0, 101 or 102.  0 indicates success, 101 indicates an error (e.g. a panic)
+and 102 indicates the node should be downgraded to a previous version.  These exit codes are read and reacted to by the
+launcher.
+
+On startup, the launcher either tries to read its previously cached state from disk, or assumes a fresh start.
+Information cached includes which mode of operation the node was executed in, the protocol version, the path to the node
+binary and the path to the config folder.
+
+On a fresh start, the launcher searches for the lowest installed version of `casper-node` and starts running it in
+validator mode.
```

Comment:
This behavior is a bit confusing for newcomers who want to run a node.

It really only makes sense at genesis.  Since genesis has already occurred, we should consider bundling a trusted hash with fresh installs.  Perhaps in the long run we could do something else if this is judged to be a non-theoretical security hazard.

In that case it would start in *joiner* mode.

Reply:
Bundling a trusted hash for convenience makes sense to me, but is outside of the scope of this CEP I think.